File: Java/Problem1.java
package Java;

public class Problem1 {

    public int max(int a, int b) {
        return a > b ? a : b;
    }

    public int min(int a, int b) {
        return a < b ? a : b;
    }

    public int maxProfit(int[] prices) {
        int n = prices.length;

        if (n < 2) {
            return 0;
        }

        int[] maxProfitLeft = new int[n];
        for (int i = 0; i < n; i++) {
            maxProfitLeft[i] = 0;
        }
        int minVal = prices[0];

        for (int i = 1; i < n; i++) {
            minVal = min(minVal, prices[i]);
            maxProfitLeft[i] = max(maxProfitLeft[i - 1], (prices[i] - minVal));
        }

        int[] maxProfitRigth = new int[n];
        for (int i = 0; i < n; i++) {
            maxProfitRigth[i] = 0;
        }
        int maxVal = prices[n - 1];

        for (int i = n - 1; i > 0; i--) {
            maxVal = max(maxVal, prices[i]);
            maxProfitRigth[i] = max(maxProfitRigth[i - 1], (maxVal - prices[i]));
        }

        int bestProfit = 0;

        for (int i = 0; i < n; i++) {
            bestProfit = max(bestProfit, maxProfitLeft[i] + maxProfitRigth[i]);
        }

        return bestProfit;
    }
}

File: Python/Problem1.py
from typing import List


def maxProfit(prices: List[int]) -> int:
    n: int = len(prices)

    if n < 2:
        return 0

    max_profit_left: List[int] = [0] * n  # [0,0,0,0,0... n]
    min_val: int = prices[0]

    # best time to buy
    for i in range(1, n, 1):
        min_val = min(min_val, prices[i])
        max_profit_left[i] = max(max_profit_left[i - 1], (prices[i] - min_val))

    # best time to sell
    max_profit_rigth: List[int] = [0] * n
    max_val: int = prices[-1]
    for i in range(n - 1, 0, -1):
        max_val = max(max_val, prices[i])
        max_profit_rigth[i] = max(max_profit_rigth[i - 1], (max_val - prices[i]))

    best_profit: int = 0
    for i in range(n):
        best_profit = max(best_profit, (max_profit_left[i] + max_profit_rigth[i]))

    return best_profit


p = [3, 3, 5, 0, 0, 3, 1, 4]
print(maxProfit(p))
